### ISSUE TITLE
support mouse with forward and back buttons

### ISSUE THREAD

**Issue by okhowang** [TigerVNC/tigervnc#919]
**Is your feature request related to a problem? Please describe.**
I want use forward/back button on mouse over vnc

**Additional context**
[https://askubuntu.com/questions/236529/binding-back-forward-to-mouse-buttons](https://askubuntu.com/questions/236529/binding-back-forward-to-mouse-buttons)

**Comment by samhed** [comment]
Agreed! This would be nice.

The VNC protocol currently only uses one byte for the state of the mouse buttons when talking to the server. Each bit indicates if a mouse button is pressed or not:

    1 left mouse button
    2 scroll wheel press
    3 right mouse button
    4 scroll up
    5 scroll down
    6 scroll left
    7 scroll right

This means there is only really space for one more button in the protocol the way it looks right now.

**Comment by okhowang** [comment]
in [rfb proto](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#gii-pseudo-encoding), there is `gii Pseudo-encoding`?
can it resolve this issue?

**Comment by CendioOssman** [comment]
Possibly. It's a bit underdocumented as it refers to the dead GGI project. I'm also not sure if there are any other implementations to do interoperability testing against.

**Comment by mdevaev** [comment]
I wanted to write the patch myself, but it looks like FLTK doesn't support additional buttons.

**Comment by mdevaev** [comment]
@CendioOssman I am making a VNC server for my [IP-KVM](https://github.com/pikvm/pikvm) and my users are asking to add additional mouse buttons. But as you've noticed, gii seems dead. Are there any other ways to get buttons?

**Comment by darkphoenix** [comment]
Would also be interested in this feature, especially if it made its way downstream to ThinLinc, which I understand is based on TigerVNC?

**Comment by samhed** [comment]
> ThinLinc, which I understand is based on TigerVNC?

Correct

**Comment by guzzard** [comment]
Would it be a massive task to add client side mouse/keyboard mapping to work around this issue? Similar to how it's possible to map mouse/keyboard buttons on x11vnc for example, but on the client side instead of the server side.

-buttonmap 89-:Alt_L+Left::Alt_L+Right

And with that mapping send keyboard keys to vnc server for backward/forward navigation in web browers, file browsers etc..

**Comment by samhed** [comment]
Note that sending the multimedia keys XF86Back and XF86Forward are not the same thing as sending these mouse buttons.

How would such a mapping be configured? Would it simply be a toggle for sending XF86Back and XF86Forward when mouse-back and forward are pressed?

**Comment by guzzard** [comment]
Yes, that is true. Above workaround is for specific use cases where XF86Back and XF86Forward work as substitutes for the real mouse buttons. This might not always be the case but it works fine in web browsers, file browsers etc.

Your question got me thinking though so I had a second look at it and found a way to actually get the server to receive fake mouse clicks. It's requires some steps but it's possible and with this workaround it should work with any application on the server. The applications would interpret the input as actual mouse clicks.

Below is the setup i tested for the mouse back button.

1. Install and configure evrouter.
2. In the configuration map the mouse back button to a button that is never used on the server. In my case i chose "XF86Launch9" since this is a button that does not exist on the keyboard, is not mapped/used, and will be disregarded by OS/applications since it's not used.
```
Window "OtherComputer:0 - TigerVNC": # Window title
"Logitech Gaming Mouse G400" "/dev/input/by-id/usb-Logitech_Gaming_Mouse_G400-event-mouse" none key/275 "XKey/XF86Launch9"
```
3. Start evrouter on the client
4. On the server side install `xbindkeys` and `xte`.
5. In `~/.xbindkeysrc` configure so that XF86Launch9 is mapped to mouse button 8
```
"xte 'mouseclick 8'"
    XF86Launch9
```
6. Run xbindkeys on the server `xbindkeys -f ~/.xbindkeysrc`

With this setup evrouter sends keyboard key XF86Launch9 to the VNC server when mouse button 8 is clicked, and the server captures the key and maps it back to mouse button 8 click.

I tested and it works great everywhere. Also, XF86Launch9 is not passed on since its captured and remapped by xte. So you just need to pick to keyboard keys which are never used anyway, and use them for mouse 8/9 instead. `# xmodmap -pk` list all keys, so should be easy to find keys that are unused normally, XF86Launch1-9 are probably good options for most people.

**Comment by samhed** [comment]
Cool workaround :+1: However, implementing something like directly in TigerVNC would likely take similar time and effort as solving it 'properly' using some sort of extension.

**Comment by guzzard** [comment]
@samhed proper extension is definitely the way to go :+1:. But now at least I have working mouse buttons until that is implemented :slightly_smiling_face:

**Comment by hongwen000** [comment]
I implemented a workaround with autohotkey, supposing that Alt-Left/Right triggers go-back/forward in the specific application.

```
SendMode Event
#IfWinActive ahk_exe vncviewer.exe
~XButton1:: SendInput, !{Left} 
~XButton2:: SendInput, !{Right}
return
```

**Comment by PenaPP** [comment]
Is there any update on this? Workarounds mentioned on https://wiki.archlinux.org/title/TigerVNC#Workaround_for_mouse_back_and_forward_buttons_not_working seem to be usable for unix clients only.

**Comment by CendioOssman** [comment]
Nope. I'm not aware of anyone working on this.

**Comment by CendioOssman** [comment]
Fixed in #1826.
